swap/peer.go

```diff
@@ -137,3 +123,93 @@ func (sp *Peer) handleErrorMsg(ctx context.Context, msg interface{}) error {
 	// maybe balance disagreement
 	return nil
 }
+
+// processAndVerifyCheque verifies the cheque and compares it with the last received cheque
+// if the cheque is valid it will also be saved as the new last cheque
+func (sp *Peer) processAndVerifyCheque(cheque *Cheque) error {
+	if err := sp.verifyChequeProperties(cheque); err != nil {
+		return err
+	}
+
+	lastCheque := sp.loadLastReceivedCheque()
+
+	// TODO: there should probably be a lock here?
+	expectedAmount, err := sp.swap.oracle.GetPrice(cheque.Honey)
+	if err != nil {
+		return err
+	}
+
+	if err := sp.verifyChequeAgainstLast(cheque, lastCheque, expectedAmount); err != nil {
+		return err
+	}
+
+	if err := sp.saveLastReceivedCheque(cheque); err != nil {
+		log.Error("error while saving last received cheque", "peer", sp.ID().String(), "err", err.Error())
+		// TODO: what do we do here?
+	}
+
+	return nil
+}
+
+// verifyChequeProperties verifies the signautre and if the cheque fields are appropiate for this peer
+// it does not verify anything that requires knowing the previous cheque
+func (sp *Peer) verifyChequeProperties(cheque *Cheque) error {
+	if cheque.Contract != sp.contractAddress {
+		return fmt.Errorf("wrong cheque parameters: expected contract: %x, was: %x", sp.contractAddress, cheque.Contract)
+	}
+
+	// the beneficiary is the owner of the counterparty swap contract
+	if err := sp.swap.verifyChequeSig(cheque, sp.beneficiary); err != nil {
+		return err
+	}
+
+	if cheque.Beneficiary != sp.swap.owner.address {
+		return fmt.Errorf("wrong cheque parameters: expected beneficiary: %x, was: %x", sp.swap.owner.address, cheque.Beneficiary)
+	}
+
+	if cheque.Timeout != 0 {
+		return fmt.Errorf("wrong cheque parameters: expected timeout to be 0, was: %d", cheque.Timeout)
+	}
+
+	return nil
+}
+
+// verifyChequeAgainstLast verifies that serial and amount are higher than in the previous cheque
+// furthermore it cheques that the increase in amount is as expected
+func (sp *Peer) verifyChequeAgainstLast(cheque *Cheque, lastCheque *Cheque, expectedAmount uint64) error {
+	actualAmount := cheque.Amount
+
+	if lastCheque != nil {
+		if cheque.Serial <= lastCheque.Serial {
+			return fmt.Errorf("wrong cheque parameters: expected serial larger than %d, was: %d", lastCheque.Serial, cheque.Serial)
+		}
+
+		if cheque.Amount <= lastCheque.Amount {
+			return fmt.Errorf("wrong cheque parameters: expected amount larger than %d, was: %d", lastCheque.Amount, cheque.Amount)
+		}
+
+		actualAmount -= lastCheque.Amount
+	}
+
+	// TODO: maybe allow some range around expectedAmount?
+	if expectedAmount != actualAmount {
+		return fmt.Errorf("unexpected exchange rate used for honey, expected %d was %d", expectedAmount, actualAmount)
```

Comment:
I'm not sure I understand this error message. The exchange rate could have been different since the expected amount differs from the actual amount, right?

But then you are showing the amounts in the error message, not the exchange rates, no?

Reply:
True, how about `fmt.Errorf("unexpected amount for honey, expected %d was %d", expectedAmount, actualAmount)`?

Reply:
Sounds good to me!